File: impls/ts/printer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pr_str = function (data, print_readability) {
    if (print_readability === void 0) { print_readability = true; }
    if (!data) {
        return "";
    }
    if (data.type === "atom") {
        return "(atom " + exports.pr_str(data.value, print_readability) + ")";
    }
    if (data.type === "keyword") {
        return data.value.slice(1);
    }
    if (data.type === "string") {
        if (print_readability) {
            var str = data.value.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\n/g, "\\n");
            return "\"" + str + "\"";
        }
        return data.value;
    }
    if (data.type === "number") {
        return data.value.toString();
    }
    if (data.type === 'error') {
        return exports.pr_str(data.value);
    }
    if (data.type === "symbol") {
        return data.value;
    }
    if (data.type === "nil") {
        return "nil";
    }
    if (data.type === "function") {
        return "#<function>";
    }
    if (data.type === "bool") {
        return data.value.toString();
    }
    if (data.type === "map") {
        var content = Object.entries(data.value).reduce(function (acc, _a, i, src) {
            var key = _a[0], value = _a[1];
            var printedKey = key.startsWith("ʞ") ?
                key.slice(1) : "\"" + key + "\"";
            acc += printedKey;
            acc += " ";
            acc += exports.pr_str(value, print_readability);
            if (i < src.length - 1) {
                acc += ' ';
            }
            return acc;
        }, "");
        return "{" + content + "}";
    }
    if (data.type === "list") {
        return "(" + data.value
            .map(function (val) { return exports.pr_str(val, print_readability); })
            .join(" ") + ")";
    }
    if (data.type === "vector") {
        return "[" + data.value
            .map(function (val) { return exports.pr_str(val, print_readability); })
            .join(" ") + "]";
    }
    return "";
};

File: impls/python/printer.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
def pr_str(val, print_readability = True):
    if val.type == "nil":
        return "nil"
    if val.type == "bool":
        if val.value:
            return "true"
        return "false"
    if (val.type == "number"):
        return str(val.value)
    if (val.type == "keyword"):
        return val.value.replace("ʞ", "")
    if (val.type == "symbol"):
        return val.value
    if (val.type == "string"):
        if not print_readability:
            return val.value
        res = ""
        for i in val.value:
            if i == '"':
                res += '\\"'
            elif i == '\\':
                res += '\\\\'
            elif i == '\n':
                res += '\\n'
            else:
                res += i
        return '"' + res + '"'
    if (val.type == "list"):
        content = []
        for i in val.value:
            content.append(pr_str(i, print_readability))
        return "(" + " ".join(content) + ")"

    if (val.type == "vector"):
        content = []
        for i in val.value:
            content.append(pr_str(i, print_readability))
        return "[" + " ".join(content) + "]"

    if (val.type == "hashmap"):
        content = []
        for key in val.value:
            if key[0:2] == "ʞ":
                content.append(key[2:])
            else:
                content.append('"' + key + '"')
            content.append(pr_str(val.value[key], print_readability))
        return "{" + " ".join(content) + "}"

    if (val.type == "error"):
        return "Error: " + pr_str(val.value, print_readability)

    if (val.type == "fn" or val.type == "custom_fn"):
        return "#function"

    if val.type == "atom":
        return "(atom " + pr_str(val.value, print_readability) + ")"

